File: tools/transferProbe.py
# { "Depends": "py-genlayer:1jb45aa8ynh2a9c9xn3b7qqh8sm5q93hwfp7jqmwsfhh8jpz09h6" }
"""Minimal probe for the withdraw() defect: how do you pay an EOA from GenVM?

AiPet.withdraw() reaches FINALIZED and moves nothing; consensus logs
UnissuedMessagesAtFinalization(unissuedCount=1). Nothing about the pet matters here,
so this strips the transfer down to its own contract and A/Bs the routes.

Measured on Testnet Bradbury (see ROADMAP §3b):

    pay_on_finalized -> EOA        ❌ message type 1, data 0xc20680 — unissued
    pay_on_accepted  -> EOA        ❌ same message — never processed
    pay_to           -> contract   ✅ paid instantly, 0.01 GEN moved
    eth_pay_owner    -> EOA        message type 0, data 0x — the right shape

The difference is the calldata. emit_transfer on a GenVM proxy always posts a
__receive__ *method call*, and an externally owned account has no code to run it.
gl.evm.contract_interface posts an EthSend with empty calldata instead — a plain
value transfer. Run with: node tools/transferProbe.mjs
"""

from genlayer import *


@gl.evm.contract_interface
class _Payee:
    """An EVM-side proxy. Its emit_transfer issues EthSend with EMPTY calldata —
    a plain value transfer, not a __receive__ call — which is the only shape an
    externally owned account can accept."""

    class View:
        pass

    class Write:
        pass


class Vault(gl.Contract):
    owner: Address
    received: u256

    def __init__(self) -> None:
        self.owner = gl.message.sender_address
        self.received = u256(0)


    @gl.public.write.payable
    def fund(self) -> None:
        """Put money in. Nothing else."""

    @gl.public.write.payable
    def __receive__(self) -> None:
        """Accept plain transfers, so a contract recipient is a fair test."""
        self.received = u256(int(self.received) + int(gl.message.value))

    @gl.public.write
    def pay_to(self, recipient: str, amount_wei: int) -> None:
        gl.get_contract_at(Address(recipient)).emit_transfer(
            value=u256(int(amount_wei)), on="accepted"
        )

    @gl.public.write
    def eth_pay_owner(self, amount_wei: int) -> None:
        """The candidate fix for AiPet.withdraw()."""
        _Payee(self.owner).emit_transfer(value=u256(int(amount_wei)))

    @gl.public.write
    def pay_on_finalized(self, amount_wei: int) -> None:
        gl.get_contract_at(self.owner).emit_transfer(value=u256(int(amount_wei)))

    @gl.public.write
    def pay_on_accepted(self, amount_wei: int) -> None:
        gl.get_contract_at(self.owner).emit_transfer(
            value=u256(int(amount_wei)), on="accepted"
        )

    @gl.public.view
    def held(self) -> str:
        return str(int(self.balance))

    @gl.public.view
    def who(self) -> str:
        return self.owner.as_hex

File: tools/transferProbe.mjs
/**
 * Does gl.evm.contract_interface(...).emit_transfer() pay an EOA?
 * This is the candidate fix for AiPet.withdraw().
 *
 *   node tools/transferProbe.mjs
 */
import { readFileSync } from "fs";
import { createClient, createAccount } from "genlayer-js";
import { testnetBradbury } from "genlayer-js/chains";
import { installGasBuffer } from "../lib/gasBuffer.mjs";
import { waitForSettled, statusName } from "../lib/waitForSettled.mjs";

installGasBuffer();

const pk = readFileSync(".env", "utf8").match(
  /^\s*[A-Z_0-9]*PRIVATE_KEY[A-Z_0-9]*\s*=\s*(0x[0-9a-fA-F]{64})/m,
)?.[1];
const account = createAccount(pk);
const client = createClient({ chain: testnetBradbury, account });

const FUND = 20_000_000_000_000_000n;
const PAY = 10_000_000_000_000_000n;

const tx0 = await waitForSettled(
  client,
  await client.deployContract({ code: new Uint8Array(readFileSync("tools/transferProbe.py")), args: [] }),
);
const address = tx0.recipient;
console.log(`vault ${address} (${statusName(tx0.status)})`);

const bal = async (a) => (await client.getBalance({ address: a })).toString();

async function send(fn, args = [], value = 0n) {
  const tx = await waitForSettled(client, await client.writeContract({ address, functionName: fn, args, value }));
  console.log(`  ${fn}(${args}) → ${statusName(tx.status)} / ${tx.txExecutionResultName ?? "?"}`);
  const msgs = JSON.stringify(tx.messages ?? [], (k, v) => (typeof v === "bigint" ? v.toString() : v));
  if (msgs !== "[]") console.log(`    message: ${msgs}`);
  return tx;
}

await send("fund", [], FUND);
const ownerBefore = BigInt(await bal(account.address));
console.log(`vault ${await bal(address)} | owner ${ownerBefore}`);

console.log("\n>>> eth_pay_owner — EthSend with empty calldata, straight to an EOA");
await send("eth_pay_owner", [PAY]);

for (const wait of [0, 15, 30]) {
  if (wait) await new Promise((r) => setTimeout(r, 15_000));
  console.log(`  +${wait}s  vault ${await bal(address)}  owner ${await bal(account.address)}`);
}

const moved = BigInt(await bal(address)) < FUND;
console.log(`\nverdict: ${moved ? "✅ an EOA CAN be paid this way — this is the withdraw() fix" : "❌ still not delivered"}`);
console.log(`vault: ${address}`);
